frontend/catalyst/device/qjit_device.py

```diff
@@ -216,12 +291,18 @@ def default_expand_fn(self, circuit, max_expansion=10):
             circuit: circuit to expand
             max_expansion: the maximum number of expansion steps if no fixed-point is reached.
         """
+
+        max_expansion = (
+            self.original_device.max_expansion
+            if hasattr(self.original_device, "max_expansion")
+            else 10
+        )
```

Comment:
No, it is stored in the old device api's device as a default argument. See [here](https://github.com/PennyLaneAI/catalyst/blob/5cc12399d30bd42ae15b11ab12878030ad5fb9ee/frontend/catalyst/device/qjit_device.py#L202). Here I just preserved the logic.

Reply:
No, it's not stored in the old device API. This is not a device attribute. 

It can be passed to the `QNode`, which will, for the old device API, pass it along to the `default_expand_fn`, but it's never stored as `device.max_expansion`. That means this will always default to 10, no matter what you put in when you call `default_expand_fn`.

At any rate, what does it have to do with program verification?

Reply:
Oh, you are right in that it would ignore the argument `max_expansion`. I fixed it for now by checking it as well.

>This is not a device attribute
I set it in tests [here](https://github.com/PennyLaneAI/catalyst/blob/a666361813067c992e40d3e4fe531d95ace1ff39/frontend/test/pytest/test_verification.py#L74)

> At any rate, what does it have to do with program verification?
I set `max_expansion=0` to disable the decomposition. In other words, I need some way to write a custom device doing bad or no decomposition to emulate failures. Let me know if you have a better suggestion, I will be glad to set it.

Reply:
>I set it in tests [here](https://github.com/PennyLaneAI/catalyst/blob/a666361813067c992e40d3e4fe531d95ace1ff39/frontend/test/pytest/test_verification.py#L74)

You _create_ it in tests there. It's not part of the API, and none of the "real" devices have that as an attribute. I don't think we should add things to source code just to accommodate test devices in ways that depart from the standard API.

What if we unit test by directly creating valid and invalid tapes and passing them to each validation function we include in preprocess? As I understand it, the changes made in [this PR](https://github.com/PennyLaneAI/catalyst/pull/711) recently should allow us to create a tape with nested `HybridOps` in it directly, outside of any queuing context.

Reply:
> You create it in tests there

Yes, I do this for the `CustomDevice`. I think it is perfectly fine. 

> It's not part of the API

A really questionable proposal is that I _read_ this attribute, but again, I do it in our custom `QJITDevice`. Basically, I need to pass a parameter to a transformation - and see no good way of doing it. Other options that I considered 1) Add this parameter to the `preprocess` - but this would break api for real; 2) pass this option through the constructor of the QJITDevice - but unfortunately we create this device in the middle of unrelated pipeline. So I found that making it like this  
might be a compromise.

> What if we unit test by directly creating valid and invalid tapes and passing them to each validation function we include in preprocess?

This way we would only test the fact that verification is defined somewhere and work correctly. I would like to also test that the verification really happens in main usage scenarios.

Reply:
Hi folks, just catching up on this discussion. From what I can see of this scenario, we should avoid appending new entities to types that migrate from tests to the codebase --- in this instance, creating something on the device is fine to use within the test infrastructure, but expecting that to exist in the main codebase is not. In this scenario, given that everything is ready validated and tested from the main codebase, we should trust it continues to work here. From the technical perspective, programming to APIs provided to us should be valid if they are well tested already.  We should commit to not making this test-specific addition right now, and move ahead with this.

If there is still want to do additional integration testing, I will suggest this happens in a follow up PR (@grwlf  feel free to create a story in the backlog for this), and to use a mocking-based solution where one can explicitly control what the broken input to the validation should look like, and check it is caught. Sound reasonable?

Reply:
Can we not control the max_expansion from a test via the [QNode](https://docs.pennylane.ai/en/stable/code/api/pennylane.qnode.html) parameter?

Reply:
The new device API ignores `max_expansion` from the qnode call. Allowing users a shortcut to skip decomposition and/or validation causes more problems than it solves. 

It still exists (for the old API, and for direct calls of the decomposition transform) because removing it is not a high priority and there's potentially a bit of legacy code to deal with, but it should eventually be gone completely.

Reply:
Set it up with a mocking `patch` so we can keep the integration tests without the `max_expansion` changes 🙂

Reply:
Nice!